Add enigma keymap for contra keyboard

christopherswenson opened the pull request:
This adds a fun custom keymap I made for my contra keyboard, which can emulate an Enigma machine.

This also makes a small change to the contra keyboard `config.h` file, to resolve a redefinition issue with RBG under-glow.

## Description

### Enigma Keymap

By default, this layout functions like a normal QWERTY layout. But it also has an in-built Enigma machine emulator. To enable the emulator, press `Fn+.`, and to go back to QWERTY, press `FN+,`.

A demonstration video can be found [here](https://youtu.be/p8kBjP1DCzo).

When Enigma mode is enabled, all letters A-Z will be enciphered using the Enigma emulator. Other keycodes will not be modified. Shift will be sent as normal, so you can choose whether to capitalize the output. If any other modifier (Ctrl, Alt, or Gui) are being pressed, the normal letter will come through.

There are a few key combinations for configuring and diagnosing the Enigma emulator:

* `Fn+Q` resets the emulator to the last-configured settings
* `Fn+W` types out the current Enigma settings, e.g. `B. IA-IIA-IIIA (A, A, A) QW` (meaning reflector B; rotors I, II, and III all at ring setting A; all at position A, and with Q and W swapped on the plugboard)
* `Fn+P` types out 1000 enciphered As, for testing
* `Fn+Backspace` back-rotates the rotors one step, for fixing typos without having to re-type a whole message
* `Fn+Z` followed by A, B, or C is used to set the reflector
* `Fn+X` followed by three characters A-E is used to set the rotors
* `Fn+C` followed by three characters A-Z is used to set the rotor positions
* `Fn+V` followed by three characters A-Z is used to set the ring settings
* `Fn+B` followed by up to 13 pairs of characters A-Z (one at a time, not simulaneously) then the Enter key is used to set the plugboard

### Change to Contra

Contra's `config.h` has this in it:

```
#ifdef RGB_DI_PIN
#define RGBLIGHT_ANIMATIONS
#define RGBLED_NUM 0
#define RGBLIGHT_HUE_STEP 8
#define RGBLIGHT_SAT_STEP 8
#define RGBLIGHT_VAL_STEP 8
#endif
```

And if a keymap attempts to define any of these values differently, compilation fails due to the value being redefined. 

This PR simply removes this block entirely. This is OK because only one other keymap for contra (`maxr1998`) actually defines `RGB_DI_PIN`, and it also attempts to redefine `RGBLIGHT_VAL_STEP` and `RGBLED_NUM`. So this PR actually fixes that keymap, too.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**]
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented on keyboards/contra/keymaps/enigma/keymap.c:
```diff
@@ -0,0 +1,708 @@
+    ),
+};
+
+void send_character(char letter) {
```
You could probably use `send_char` for this instead.

drashna commented on keyboards/contra/keymaps/enigma/keymap.c:
```diff
@@ -0,0 +1,708 @@
+                  return save_setting();
+              }
+              break;
+          case EN_A:
```
This entire block could probably be replaced with just:
```c
    case EN_A ... EN_Z: {
        uint8_t index = keycode - EN_A;
        normal_keycode = KC_A + index;
        letter = 'A' + index;
        break;
        }
```

drashna commented on keyboards/contra/keymaps/enigma/keymap.c:
```diff
@@ -0,0 +1,708 @@
+              break;
+      }
+    }
+    int mods = get_mods();
+    int any_mods = (mods & MOD_MASK_CTRL) || (mods & MOD_MASK_ALT) || (mods & MOD_MASK_GUI);
```
get_mods is a "uint8_t" type
```suggestion
    uint8_t mods = get_mods();
    bool any_mods = (mods & MOD_MASK_CTRL) || (mods & MOD_MASK_ALT) || (mods & MOD_MASK_GUI);
```

drashna commented on keyboards/contra/keymaps/enigma/rules.mk:
```diff
@@ -0,0 +1,4 @@
+AUTO_SHIFT_ENABLE = no
+FORCE_NKRO = yes
```
This is a define, not a make rule.
```suggestion
```

christopherswenson commented:
@drashna Thanks for the great review. I made all the changes you requested. I fixed the repetition of the `case EN_x` stuff a bit differently than you suggested, but basically the same idea.

I also fixed an issue where `Fn+Backspace` was just not mapped to the rotor reverse key code... must have accidentally removed it while I was cleaning up the layouts.

Let me know if you spot anything else that I should change.

christopherswenson commented:
I'm actually currently working on a few more refactors. Didn't like how big and scary the `process_record_user` function was. I will comment back when done.

christopherswenson commented:
Okay, I think I'm done refactoring

keyboard-magpie commented:
Exceptionally cool. One to add to the news articles as a "look what QMK is capable of".

drashna commented:
You are very welcome!
> fixed the repetition of the `case EN_x` stuff a bit differently than you suggested, but basically the same idea.

And yeah, I hate to see repetitive code, especially if it can be condensed down a lot! 

And very cool code!
